pkg/apis/v1alpha1/edge-placement.go

```diff
@@ -0,0 +1,130 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v2alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +crd
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// EdgePlacement binds a collection of (a) Namespaces and non-namespaced objects
+// to (b) a collection of Locations.
+// An EdgePlacement object appears in the center and directs all the syncable
+// objects in the selected Namespaces and all the selected non-namespaced objects
+// to propagate to _all_ of the selected Locations
+// (one SyncTarget in each such Location).
+// This is not entirely unrelated to a TMC Placement, which directs the selected
+// Namespaces to propagate to _one_ of the selected Locations.
+//
+// Overlap between EdgePlacements is allowed:
+// two different EdgePlacement objects may select intersecting Location sets
+// and/or intersecting Namespace sets.
+// This is not problematic because:
+//   - propagation _into_ a destination is additive;
+//   - propagation _from_ a source is additive;
+//   - two directives to propagate the same object to the same destination are
+//     simply redundant (they, by design, can not conflict).
+type EdgePlacement struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard object metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// `spec` describes a desired binding between workload and Locations.
+	// Unlike a TMC Placement, there is an inherent multiplicity and dynamicity
+	// in the set of Locations that will be synced to and this field
+	// never shifts into immutability.
+	// +optional
+	Spec EdgePlacementSpec `json:"spec,omitempty"`
+
+	// `status` describes the status of the process of binding
+	// workload to Locations.
+	// +optional
+	Status EdgePlacementStatus `json:"status,omitempty"`
+}
+
+// EdgePlacementSpec holds a desired binding between (a) Namespaces and non-namespaced objects
+// and (b) Locations.
+type EdgePlacementSpec struct {
+	// `locationWorkspaceSelector` identifies the workspaces in which to look for Location
+	// objects, in terms of labels on the Workspace objects.
+	LocationWorkspaceSelector metav1.LabelSelector `json:"locationWorkspaceSelector,omitempty"`
+
+	// `locationSelectors` identifies the relevant Location objects in terms of their labels.
+	// A Location is relevant if and only if it passes any of the LabelSelectors in this field.
+	LocationSelectors []metav1.LabelSelector `json:"locationSelectors,omitempty"`
```

Comment:
Does this mean that first the LocationWorkspaceSelector is applied and then the LocationSelector is applied for each of the workspaces selected by the first selector ?

Reply:
Yes

Reply:
I am still learning about APIExport and binding in kcp.  I suspect that has a different way to solve the problem: some _service provider_ does APIExport of `Location` from one of their workspaces, and maintains controllers that read --- indirectly through the one workspace with the APIExport --- all the Locations objects based on APIBindings of that APIExport.

Reply:
Right, I think that would affect the design of this API.